### PR TITLE
Plasma weapon overheating

### PR DESCRIPTION
**Pull request by caelaislinn**
:cl: CaelAislinn
imageadd: Adds a visual effect for plasma weapon heat now so you can tell if you're about to overheat. 
bugfix: Plasma pistol will now instantly overheat after firing an overcharge shot. It will also consume a lot more energy to use. 
experiment: Removes chatlog spam about firing a gun. 
/:cl:

### CODE REVIEW

File: code/modules/projectiles/gun_overheat.dm [hunk 1]
```diff
@@ -0,0 +1,84 @@
+
+/obj/item/weapon/gun
+	//Fire delay is modified by this, then used to determine the time after last firing to count as the weapon
+	//recovering from all the overheat it has accumulated
+	//Example: fire_delay 10 with a fullclear_mod of 2 means the player must wait for 20 before firing for the heat to
+	//be cleared.
+	var/overheat_fullclear_delay = 3 SECONDS
+	var/overheat_fullclear_at = 0
+	var/overheat_sfx = null
+	var/overheat_capacity = -1 //If set above -1, a weapon will decrement this counter per-shot until it hits 0.
```
Comment (X0-11) [inline code/modules/projectiles/gun_overheat.dm]: Comment now inaccurate?

File: code/modules/projectiles/gun.dm [hunk 2]
```diff
@@ -505,6 +482,7 @@
 
 //called after successfully firing
 /obj/item/weapon/gun/proc/handle_post_fire(mob/user, atom/target, var/pointblank=0, var/reflex=0)
+	/*
```
Comment (X0-11) [inline code/modules/projectiles/gun.dm:485]: Ah, the nuclear option.

File: code/modules/projectiles/gun.dm [hunk 3]
```diff
@@ -373,18 +345,15 @@
...
-		return
-
 	if(stored_targ)
 		stored_targ = target
 		visible_message("<span class = 'notice'>[user] refocuses their aim on [target]</span>")
 		return
 
 	if(world.time < next_fire_time)
-		if (world.time % 3) //to prevent spam
+		/*if (world.time % 3) //to prevent spam
```
Comment (X0-11) [inline code/modules/projectiles/gun.dm:354]: Why?

File: code/modules/projectiles/gun_overheat.dm [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	var/heat_current = 0
+	var/datum/progressbar/heat_bar
+	var/heat_per_shot = 1
+
+/obj/item/weapon/gun/process()
+	world << "[src.type]/process()"
```
Comment (X0-11) [inline code/modules/projectiles/gun_overheat.dm]: debug code?

File: code/modules/projectiles/gun_overheat.dm [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+	if(process_heat())
+		return
+
+	return PROCESS_KILL
+
+/obj/item/weapon/gun/proc/process_heat()
+	world << "/obj/item/weapon/gun/proc/process_heat()"
```
Comment (X0-11) [inline code/modules/projectiles/gun_overheat.dm]: Debug code?

File: code/modules/projectiles/gun_overheat.dm [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+					color = "#ff0000"
+
+		//continue processing
+		return 1
+
+	//stop processing
+	return 0
+
+/obj/item/weapon/gun/proc/add_heat(var/new_val)
+	world << "/obj/item/weapon/gun/proc/add_heat([new_val])"
```
Comment (X0-11) [inline code/modules/projectiles/gun_overheat.dm]: ?

File: code/modules/projectiles/gun.dm (hunk 3 above)
Comment (caelaislinn) [reply code/modules/projectiles/gun.dm:354]: To prevent spam harder

Comment (X0-11) [reply code/modules/projectiles/gun.dm:354]: I think that's there to prevent message spam

Comment (caelaislinn) [reply code/modules/projectiles/gun.dm:354]: Yeah, which is why I've commented it all out.

Comment (X0-11) [reply code/modules/projectiles/gun.dm:354]: Why do we want more message log spam?

Comment (caelaislinn) [reply code/modules/projectiles/gun.dm:354]: Look at the raw file, I've commented out the entire block

**Pull request merged**